cpp/src/io/csv/reader_impl.cu

```diff
@@ -598,30 +412,224 @@ reader::impl::load_data_and_gather_row_offsets(host_span<char const> data,
+
+    // TODO: Allow parsing the header outside the mapped range
+    CUDF_EXPECTS((range_offset == 0 || reader_opts.get_header() < 0),
+                 "byte_range offset with header not supported");
+
+    // Gather row offsets
+    auto data_row_offsets =
+      load_data_and_gather_row_offsets(reader_opts,
+                                       parse_opts,
+                                       header,
+                                       h_data,
+                                       data_start_offset,
+                                       (range_size) ? range_size : h_data.size(),
+                                       (skip_rows > 0) ? skip_rows : 0,
+                                       num_rows,
+                                       load_whole_file,
+                                       stream);
+    auto& row_offsets = data_row_offsets.second;
+    // Exclude the rows that are to be skipped from the end
+    if (skip_end_rows > 0 && static_cast<size_t>(skip_end_rows) < row_offsets.size()) {
+      row_offsets.shrink(row_offsets.size() - skip_end_rows);
+    }
+    return data_row_offsets;
+  }
+  return {rmm::device_uvector<char>{0, stream}, selected_rows_offsets{stream}};
+}
+
+std::vector<data_type> get_data_types_from_column_names(
+  std::vector<column_parse::flags> const& column_flags,
+  std::map<std::string, data_type> const& column_type_map,
+  std::vector<std::string> const& column_names,
+  int32_t num_actual_columns)
+{
+  std::vector<data_type> selected_dtypes;
+
+  for (int32_t i = 0; i < num_actual_columns; i++) {
+    if (column_flags[i] & column_parse::enabled) {
+      auto const col_type_it = column_type_map.find(column_names[i]);
+      CUDF_EXPECTS(col_type_it != column_type_map.end(),
+                   "Must specify data types for all active columns");
+      selected_dtypes.emplace_back(col_type_it->second);
+    }
+  }
+
+  return selected_dtypes;
+}
+
+std::vector<data_type> select_data_types(std::vector<column_parse::flags> const& column_flags,
+                                         std::vector<data_type> const& dtypes,
+                                         int32_t num_actual_columns,
+                                         int32_t num_active_columns)
+{
+  std::vector<data_type> selected_dtypes;
+
+  if (dtypes.size() == 1) {
+    // If it's a single dtype, assign that dtype to all active columns
+    selected_dtypes.resize(num_active_columns, dtypes.front());
+  } else {
+    // If it's a list, assign dtypes to active columns in the given order
+    CUDF_EXPECTS(static_cast<int>(dtypes.size()) >= num_actual_columns,
+                 "Must specify data types for all columns");
+
+    for (int i = 0; i < num_actual_columns; i++) {
+      if (column_flags[i] & column_parse::enabled) { selected_dtypes.emplace_back(dtypes[i]); }
+    }
+  }
+  return selected_dtypes;
+}
+
+std::vector<data_type> parse_column_types(std::vector<column_parse::flags>& column_flags,
```

Comment:
Looks like https://github.com/rapidsai/cudf/pull/9079 hasn't been merged (correctly). This function should be removed since passing types as strings is no longer supported.

Reply:
I definitely handled the merge conflicts badly, then. I'll get your PR merged back in. Sorry!

Reply:
I probably clobbered the same changes for JSON in https://github.com/rapidsai/cudf/pull/9088. I'll get those cleaned up too.